### ISSUE TITLE
Behavior when functions return `nothing`

### ISSUE THREAD

**Issue by emstoudenmire** [GTorlai/Observers.jl#8]
We are in the process of piloting the use of Observers in an updated implementation of ITensor DMRG as well as TDVP. One practical issue I just encountered when testing it is how to handle cases where on a given step I don't want to do any measurement. For example, `update!` is called at every step of TDVP but I only want to record a single measurement at each time step (so at the end of one sweep).

Here is a proposed behavior change: if a measurement function returns `nothing` then this value is not appended to the list of results. I would be happy to submit a PR for this.

Here are two possible downsides to this design change though:
- the resulting Observer could now store results vectors of different lengths. This seems ok, but I wonder if it might break the conversion to DataFrames feature? 
- what if for some reason a user wants to record the actual value `nothing`? This seems easy enough to work around though, like the user could return some other stand-in value such as the string "nothing" or another default value of their choosing.

**Comment by mtfishman** [comment]
This was the plan I had originally, I thought I had already implemented it but I guess not. Anyway, I think this is the right way to go.

In terms of the first issue, it would in principle break the conversion to `DataFrames`:
```julia
julia> using DataFrames

julia> DataFrame(Dict("X" => [1, 2, 3], "Y" => [4, 5, 6]))
3×2 DataFrame
 Row │ X      Y     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      4
   2 │     2      5
   3 │     3      6

julia> DataFrame(Dict("X" => [1, 2, 3], "Y" => [4, 5]))
ERROR: DimensionMismatch("column :Y has length 2 and column :X has length 3")
Stacktrace:
 [1] DataFrame(columns::Vector{Any}, colindex::DataFrames.Index; copycols::Bool)
   @ DataFrames ~/.julia/packages/DataFrames/MA4YO/src/dataframe/dataframe.jl:192
 [2] DataFrame(d::Dict{String, Vector{Int64}}; copycols::Bool)
   @ DataFrames ~/.julia/packages/DataFrames/MA4YO/src/dataframe/dataframe.jl:286
 [3] DataFrame(d::Dict{String, Vector{Int64}})
   @ DataFrames ~/.julia/packages/DataFrames/MA4YO/src/dataframe/dataframe.jl:276
 [4] top-level scope
   @ REPL[11]:1
```
However, you'd have to wonder if that was in fact something people really wanted, because it seems like the reason to put something in a table like that would be to compare data across iterations, sweeps, etc.

In that case, you should just do some of your own processing of the data, for example if you had some data that was measured once every sweep and some data that was measured at each site then you should split that data into two parts that each fit into separate DataFrames yourself, or make them the same length and indicate that some data is missing.

As for the issue with `nothing`, I was also concerned about that, but I think we can tell people to use something else to indicate missing data, which is exactly what the `missing` type is for:
```julia
help?> missing
search: missing missings Missing Missings MissingException ismissing nonmissingtype skipmissing passmissing

  missing

  The singleton instance of type Missing representing a missing value.

  See also: NaN, skipmissing, nonmissingtype.
```
in contrast to `nothing`:
```julia
help?> nothing
search: nothing Nothing isnothing

  nothing

  The singleton instance of type Nothing, used by convention when there is no value to return (as in a C void
  function) or when a variable or field holds no value.

  See also: isnothing, something, missing.
```
If someone really wanted values of `nothing` in their results they could always insert `missing` and then do some postprocessing. Otherwise as you suggest we could make this a keyword argument where someone could specify what value they want to use to indicate skipping recording a certain result, which defaults to `nothing`.

**Comment by emstoudenmire** [comment]
Glad the idea is in line with your plans. My suggestion would be that `nothing` means no entry in the vector is generating, whereas `missing` could be something users can put if they want to skip entries but still enter a value into the results. As you probably know, `missing` has a special role in DataFrames (https://dataframes.juliadata.org/stable/man/missing/).

Another reason `nothing` is a good skipping value is that value returned by a `return` keyword or the default return in a non-explicitly-returning Julia function like `f() = print("called f")` is `nothing`.

So overall I'm for this change (and am happy to make it) and agree that people using it might accept that they either can't use DataFrames after, or should return `missing` if they wish to use DataFrames.

**Comment by mtfishman** [comment]
Closed by #9.
